### PR TITLE
Add support for Redirect preserving method/body (HTTP 307 and 308)

### PR DESCRIPTION
**Pull request by huysentruitw**
Add support for Redirect preserving method/body (HTTP 307 and 308)

Addresses #2864

### CODE REVIEW

**Comment by Tratcher** [comment]
I'd rather make this an extension method here:
https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http.Extensions/src/ResponseExtensions.cs

In hindsight we put too many trivial methods directly onto the HttpResponse that we then have to deal with in the derived implementations.

**Comment by davidfowl** [comment]
Agreed

**Comment by huysentruitw** [comment]
Should I move the other 2 `Redirect` methods too?

**Comment by Tratcher** [comment]
No, we prefer to avoid unnecessary breaking changes.

**Comment by huysentruitw** [comment]
Moved + added a unit-test.

File: src/Http/Http.Extensions/src/ResponseExtensions.cs [hunk 1]
```diff
@@ -22,5 +23,26 @@ public static void Clear(this HttpResponse response)
...
+        /// </summary>
+        /// <param name="response">The <see cref="HttpResponse"/> to redirect.</param>
+        /// <param name="location">The URL to redirect the client to. This must be properly encoded for use in http headers where only ASCII characters are allowed.</param>
+        /// <param name="permanent"><c>True</c> if the redirect is permanent (301 or 308), otherwise <c>false</c> (302 or 307).</param>
+        /// <param name="preserveMethod"><c>True</c> if the redirect needs to reuse the method and body (308 or 307), otherwise <c>false</c> (301 or 302).</param>
+        public static void Redirect(this HttpResponse response, string location, bool permanent, bool preserveMethod)
+        {
+            if (preserveMethod)
+            {
+                response.StatusCode = permanent ? 308 : 307;
```
Comment (Tratcher) [inline src/Http/Http.Extensions/src/ResponseExtensions.cs]:
```suggestion
                response.StatusCode = permanent ? StatusCodes.Status308PermanentRedirect : StatusCodes.Status307TemporaryRedirect;
```

File: src/Http/Http.Extensions/src/ResponseExtensions.cs [hunk 2]
```diff
@@ -22,5 +23,26 @@ public static void Clear(this HttpResponse response)
...
+            }
+            else
+            {
+                response.StatusCode = permanent ? 301 : 302;
```
Comment (Tratcher) [inline src/Http/Http.Extensions/src/ResponseExtensions.cs]:
```suggestion
                response.StatusCode = permanent ? StatusCodes.Status301MovedPermanently : StatusCodes.Status302Found;
```

File: src/Http/Http.Extensions/test/ResponseExtensionTests.cs [hunk 3]
```diff
@@ -35,6 +38,23 @@ public void Clear_AlreadyStarted_Throws()
...
+        [Theory]
+        [InlineData(true, false, 301)]
+        [InlineData(false, false, 302)]
+        [InlineData(true, true, 308)]
+        [InlineData(false, true, 307)]
+        public void Redirect_SetsResponseCorrectly(bool permanent, bool preserveMethod, int expectedStatusCode)
+        {
+            var location = "http://localhost/redirect";
+            var context = new DefaultHttpContext();
+            context.Response.StatusCode = (int)HttpStatusCode.OK;
```
Comment (Tratcher) [inline src/Http/Http.Extensions/test/ResponseExtensionTests.cs]:
```suggestion
            context.Response.StatusCode = StatusCodes.Status200OK;
```

**Review by Tratcher — APPROVED** [review]

**Comment by huysentruitw** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=180954

**Comment by anurse** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=180954

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2267
https://github.com/aspnet/AspNetCore-Internal/issues/2181

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**